### DISCUSSION TITLE
What makes the OS routines bloated?

### DISCUSSION THREAD

**Discussion by tschak909** [AmigaPorts/ACE#229] [General]
Hello. I'm Thom Cherryhomes, a developer with tons of experience, not just on Amiga, but on dozens of other platforms.

I am in the process of making a game (a version of PAC-MAN, using graphics and sound data from the arcade ROMs) which uses the Amiga operating system (Exec, graphics, layers, input, audio, etc.) to implement the game, while providing source code and documentation to serve as a canonical example of how to make a game that is friendly with the system and can co-exist along with other running applications, as well as running on all possible Amiga systems and peripheral configurations.

I have not only all of the Rom Kernel Manual (RKM) documentation, but also source code for version 40 of the operating system, and am finding myself scratching my head, because what I see is neither (1) slow nor (2) bloated. I see a system that goes to great lengths to multiplex resources to be shared amongst many tasks, providing low level access to the hardware in a way that can coexist with such multiplexing.

What were your findings? Do you have any examples as to what gave you your conclusions?

-Thom

**Comment by tehKaiN** [comment]
Hi, sorry for late reply, I just reached my PC.

I've started coding up ACE 11 years ago, with intent of having games working under kickstart 1.3 and stock configuration of A500. The first version of ACE was OS-friendly, basically just a wrapper extending graphics.library's functionality. I've found some blockers which made me ultimately drop using the graphics.library:

- If you look at [documentation of ScrollVPort](http://amigadev.elowar.com/read/ADCD_2.1/Includes_and_Autodocs_3._guide/node031C.html), you will see that there's a BUGS section which states that under ks1.3 the function is 10x slower than it should be. It's a common problem that ks1.3 versions of functions are either bugged or non-existent. Lots of convenient stuff was added in version 2.0+, some of it in 3.0+ (think lowlevel.library)
- Later on, I tried to implement [Scrolling Tricks](http://megaburken.net/~patrik/ScrollingTricks/ScrollingTricks/Docs/index-uk.html), mainly the semi-infinite and memory-efficient Y scroll which would wrap around the bitmap using copper, and I've found that doing so with OS-friendly UCopLists isn't the best idea - it runs next to copper instructions added by OS internals, which gave me a hard time when trying to set the bitplane pointers to required values because OS kept overriding that.
- Even after I've semi-fixed that, after adding the memory-efficient corkscrew X scroll, the viewport/rastport abstraction became very limiting, because of built-in layers thingy - my blits done out of legal bitmap bounds became cropped even though they should work fine due to how the corkscrew stuff works, and there was no way of disabling this safety mechanism. One might argue that it's only present in rastport-aware blit functions, but there is no function which allows for masked blits and that wouldn't use the rastport at the same time.
- Also, no interleaved bitmap support on ks1.3, which ACE supports without problems
- Also you can't do any assumptions about performance - you don't know what patches/utilities ppl have installed in their OSes, so you can't assume that OS will take certain percentage of CPU/memory resources. Also the CPU overhead of the OS is very much variable - from my tests the clean OS can eat up anything from 10 to 90% of frame time, and it can heavily vary from frame to frame.

I guess some things could be done differently (Amiga coding wasn't and probably still isn't documented extensively) but I've ran into performance/capability blockers so I just went with full custom code. The main takeaway is that if you want to write for OS, you need to aim with games that can be done with the API and no fancy tricks that bend the limits. Oh and it's crucial to use more modern OS version to get all the goodies and fixes.

Still, ACE isn't *that* os-unfriendly - it unfreezes it for memory allocation and file accesss. Also, there's an os-friendly branch which allows having the OS alive in the background for as long as you want, but the graphics etc are still done with custom code.
